app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt

```diff
@@ -428,18 +430,71 @@ class MooltipassDevice(private val device: BluetoothDevice, private var debug: I
 @ExperimentalCoroutinesApi
 class MooltipassScan {
 
+    private var context:Context? = null
+
     private fun pairedDevice(context: Context): BluetoothDevice? {
+        scheduleBlutoothStateObserver(context);
+
         val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
-        return bluetoothManager.adapter.bondedDevices.lastOrNull(::filter)
+        return bluetoothManager.adapter.bondedDevices.filter(::filter).firstOrNull()
     }
 
     fun deviceFlow(context: Context): Flow<BluetoothDevice> {
         return pairedDevice(context)
-            ?.let(::flowOf) // use paired device, if available...
-            ?: emptyFlow() // do not scan, as paired device is necessary
+                ?.let(::flowOf) // use paired device, if available...
+                ?: emptyFlow() // do not scan, as paired device is necessary
 //            ?:scanFlow(context).map(ScanResult::getDevice) // ... else scan devices
     }
 
+    private fun scheduleBlutoothStateObserver(context: Context)
+    {
+        this.context = context
+        val filter = IntentFilter()
+        filter.addAction(BluetoothDevice.ACTION_ACL_CONNECTED)
+        context.registerReceiver(broadcastReceiver, filter)
+    }
+
+    private val broadcastReceiver: BroadcastReceiver = object : BroadcastReceiver() {
+        override fun onReceive(context: Context?, intent: Intent) {
+
+            if (BluetoothDevice.ACTION_ACL_CONNECTED == intent.action) {
+                val currentDevice:BluetoothDevice? = intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE)
+                if(currentDevice?.address!!.startsWith(MAC_ADDRESS_BASE_VALUE,true))
+                {
+                    saveDevice(context,currentDevice)
+                }
+            }
+        }
+    }
+
+    private fun saveDevice(context: Context?, currentDevice:BluetoothDevice?)
+    {
+        val sharedPreference =  context?.getSharedPreferences("MOOLTIPASS_LAST_DEVICE",Context.MODE_PRIVATE)
+        var editor = sharedPreference?.edit()
+        editor?.putString("DEVICE_NAME",currentDevice?.name)
```

Comment:
Why is it necessary to store the DEVICE_NAME?

Reply:
guessing if we want to present the list at some point, given our latest firmware allows device name customization.